Rasterizing Tile Layer V2 (Is a new Pull request with my changes)

inforithmics opened the pull request:
I use in the Sample as Rendering Format Skp because this is what i intend to use by default. 
The reason why the Scaling didn't work is that the NavigateTo Event was fired before the Animation was finished. So that not the correct Data where loaded. I introduced an Action to Animation Entry so that this Action is executed on the Final Animation and so solving the Problem.

In the comment of OnNavigated to I found the following command

// Note. Instead of a delay it may also be possible to call Navigated immediately with the viewport state
// that is the result of the animation.

But I don't know mapsui well enough to fix this.

@pauldendulk 
So feel free to change this pull request if, I've done something not so good.

inforithmics commented:
I have now fixed all Bugs that I know of.

pauldendulk commented:
@inforithmics Sorry I could not find time for this so far. My focus is now the NTS port and I will continue with that until I have a MR for the full port. I am happy with the progress there.

inforithmics commented:
This pull request is not urgent and to have all planed features, I need NTS.

pauldendulk commented:
Sorry for causing conflicts. If you can fix them I will make time to coming week to review this.

inforithmics commented on Mapsui.Rendering.Skia/MapRenderer.cs:
```diff
@@ -69,30 +70,64 @@ public MapRenderer()
+                        break;
+                    case EStreamFormat.Skp:
+                        var picture = (SKPicture?)RenderToPicture(viewport, layers, background);
+                        picture?.Serialize(memoryStream);
```
This is for caching the Memorystream for example in a sqlite database

pauldendulk replied:
oh, that could be very useful. We should add this at some point.

inforithmics commented on Mapsui/Rendering/IRenderer.cs:
```diff
@@ -10,7 +10,8 @@ namespace Mapsui.Rendering
         void Render(object target, IReadOnlyViewport viewport, IEnumerable<ILayer> layers, IEnumerable<IWidget> widgets, Color? background = null);
-        MemoryStream? RenderToBitmapStream(IReadOnlyViewport viewport, IEnumerable<ILayer> layers, Color? background = null, float pixelDensity = 1);
+        MemoryStream? RenderToBitmapStream(IReadOnlyViewport viewport, IEnumerable<ILayer> layers, Color? background = null, float pixelDensity = 1, EStreamFormat streamFormat = EStreamFormat.Png);
+        object? RenderToPicture(IReadOnlyViewport viewport, IEnumerable<ILayer> layers, Color? background = null);
```
I will try to find a better solution. I will reduce the PR to a png Tile Renderer.

charlenni replied:
One of my older suggestions was to remove any handling of drawing of geometries from the default rendering process and replace it with style render. So they could replaced very easily with new and better ones. I did this for the vector tiles, which don't need any change of Mapsui.

charlenni replied:
Sorry, @inforithmics, saw your comment a little to late.

You want to catch all the drawings for a tile and convert them to a SKPicture. Is this correct? For this, I assume, you have to really change Mapsui. 

Two remarks:
1. You could get in trouble when a geometry crosses a tile border, because e.g. lines don't end at the tile border. If you then draw them only up to the border or over each other, the result could be different from the original (think about a line with a border and rounded ends outside the tile, that has ends when you cut it on a borders)
2. One part of the used time is to convert the geometry from Mapsui format to Skia format, but the second part is the time, Skia uses to upload always changing SKPaths to the graphics card. If you store the SKPath and always draw the same, it could be faster (not totally sure about this, on Skia side it is handled in this way, but I don't know if SkiaSharp does the same).

inforithmics commented on Mapsui.Rendering.Skia/PictureRenderer.cs:
```diff
@@ -12,12 +16,22 @@ internal class PictureRenderer
             var scaleY = rect.Height / picture.CullRect.Height;
-            var matrix = SKMatrix.CreateTranslation(rect.Left, rect.Top);
-            matrix.PostConcat(SKMatrix.CreateScale(scaleX, scaleY));
+            var matrix = SKMatrix.CreateTranslation(rect.Left / scaleX, rect.Top / scaleY);
```
Yes the scaling didn't work I found out because the tiles are scaled for a short time.

charlenni commented on Mapsui/Animations/AnimationEntry.cs:
```diff
@@ -7,13 +7,15 @@ public class AnimationEntry<T>
         private readonly double _animationDelta;
         private readonly Action<T, AnimationEntry<T>, double>? _tick;
         private readonly Action<T, AnimationEntry<T>>? _final;
+        private readonly Action? _action;
```
I'm with @pauldendulk. Animation has no special cases and could animate anything.

inforithmics replied:
This is now removed

charlenni commented:
@inforithmics It is really easy to use own style renderers. Much easier as to change Mapsui core parts ;-)

You could find an implementation in the vector tile layers (see #1478), where I use own style renderers for raster tiles, that have a special paint (could be transparent). See for this the [RasterTileStyle](https://github.com/charlenni/VectorTileLayers/blob/main/Mapsui.VectorTileLayers.Core/Styles/RasterTileStyle.cs) and the corresponding [RasterTileStyleRenderer](https://github.com/charlenni/VectorTileLayers/blob/main/Mapsui.VectorTileLayers.Core/Renderer/RasterTileStyleRenderer.cs). The style could be an empty class, that only should derive from IStyle. The renderer must implement ISkiaStyleRenderer. In the renderer than you draw the feature like you want to draw it. Before you could use it, you have to register the renderer for the style

    mapControl.Renderer.StyleRenderers[typeof(RasterTileStyle)] = new RasterTileStyleRenderer();

That's it. Now Mapsui always uses, if a feature or layer has this style, the given renderer by calling the

    public bool Draw(SKCanvas canvas, IReadOnlyViewport viewport, ILayer layer, IFeature feature, IStyle style, ISymbolCache symbolCache)

function of the renderer.

At the end, all predefined renderer should use this pattern, so that you could easily replace a renderer by your own and have not to change Mapsui core functions. The only difference to the above would be, that the default renderers are set by Mapsui while the ones you define must be set by you.

inforithmics commented:
I think I have to rearchitect things here. And I found cases where the Tiling Layer does not work. So again (WIP)

charlenni commented:
Thought about this a bit longer. 

If it is right, what I wrote above, then you try to make a first-rendering-then-drawing process by a special layer. You convert all Mapsui geometries into Skia paths/commands, that then could be drawn on the screen. The advantage is, that you don't need to convert the geometries into Skia each time you want to draw them. The disadvantage is, that you don't have the information, which geometries should be drawn on the screen, because you don't know which part of the map is shown on the screen.

One solution would be to think about the whole map is a big canvas and use this to convert the geometries into Skia paths. When drawing, you calculate the origin and scale, move and scale the canvas and draw your already rendered geometries. But that would not work, because Skia use only floats for matrices and the error is on higher zoom levels to big.

Up to now, I don't have a good solution for this, because it should work with an unknown layer. Perhaps a solution with different zoom levels would work (create Skia paths/commands for each zoom level and different origins). Or calculate the matrices by yourself and transfer them later to Skia when you know something about the visible part of the map. Or anything else?

inforithmics commented:
I fixed the Country Shape Tiling Smaple the reason it didn't work was different CRS because. Rasterizing Tiling Layer uses GlobalSphericalMercator and it now uses a IProjction to convert that.

Then I added a GeometrySimplyProvider (for simplifying Geometries) and an Sqlite PersistCache Sample. (I didn't inlcude it in Mapsui because it would add sqlite as a dependency).

pauldendulk commented:
@inforithmics Perhaps it is better to create a new PR with just the minimal changes to add tiled bitmap rasterization. The RasterizingTileSource should be the main big change, and apart from that I expect only a few changes. We can think about Skp later on.

In the meantime I going to reorganize the project structure, which will break things.

inforithmics commented:
Agreed I will make V3 of this pull request with only rasterizing tile layer and png Format
